Resource endpoint discovery is broken

originally discovered by @kikiya 

## steps

Lagom on GKE with Cassandra hosted elsewhere.

```
play.modules.enabled += "com.lightbend.rp.servicediscovery.lagom.javadsl.ServiceLocatorModule"
```

## problem

```
018-08-30 11:44:06.000 PDT
java.util.concurrent.CompletionException: java.net.URISyntaxException: Illegal character in scheme name at index 0: 128://10.128.0.6:0 at java.util.concurrent.CompletableFuture.encodeThrowable(CompletableFuture.java:292) at 
java.util.concurrent.CompletableFuture.completeThrowable(CompletableFuture.java:308) at
 java.util.concurrent.CompletableFuture.uniApply(CompletableFuture.java:593) at 
 java.util.concurrent.CompletableFuture$UniApply.tryFire(CompletableFuture.java:577) at 
 java.util.concurrent.CompletableFuture.uniApplyStage(CompletableFuture.java:617) at 
 java.util.concurrent.CompletableFuture.thenApplyAsync(CompletableFuture.java:1988) at 
 scala.concurrent.java8.FuturesConvertersImpl$CF.thenApply(FutureConvertersImpl.scala:27) at 
 scala.concurrent.java8.FuturesConvertersImpl$CF.thenApply(FutureConvertersImpl.scala:18) at 
 com.lightbend.rp.servicediscovery.javadsl.ServiceLocator.lookupOne(ServiceLocator.java:95) at 
 com.lightbend.rp.servicediscovery.lagom.javadsl.LagomServiceLocator.lambda$locate$0(LagomServiceLocator.java:44) at 
 java.util.concurrent.CompletableFuture.uniCompose(CompletableFuture.java:952) at 
 java.util.concurrent.CompletableFuture$UniCompose.tryFire(CompletableFuture.java:926) at 
 java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:474) at 
 java.util.concurrent.CompletableFuture.postFire(CompletableFuture.java:561) at 
 java.util.concurrent.CompletableFuture$UniApply.tryFire(CompletableFuture.java:580) at 
 java.util.concurrent.CompletableFuture$Completion.run(CompletableFuture.java:442) at 
 java.lang.Thread.run(Thread.java:748) Caused by: java.net.URISyntaxException: Illegal character in scheme name at index 0: 128://10.128.0.6:0 at java.net.URI$Parser.fail(URI.java:2848) at java.net.URI$Parser.checkChars(URI.java:3021) at 
 java.net.URI$Parser.checkChar(URI.java:3031) at java.net.URI$Parser.parse(URI.java:3047) at 
 java.net.URI.<init>(URI.java:673) at 
 com.lightbend.rp.servicediscovery.scaladsl.ServiceLocatorLike.$anonfun$translateResolved$1(ServiceLocator.scala:166) at 
 scala.collection.TraversableLike.$anonfun$map$1(TraversableLike.scala:234) at 
 scala.collection.Iterator.foreach(Iterator.scala:929) at 
 scala.collection.Iterator.foreach$(Iterator.scala:929) at 
 scala.collection.AbstractIterator.foreach(Iterator.scala:1417) at 
 scala.collection.IterableLike.foreach(IterableLike.scala:71) at 
 scala.collection.IterableLike.foreach$(IterableLike.scala:70) at 
 scala.collection.AbstractIterable.foreach(Iterable.scala:54) at 
 scala.collection.TraversableLike.map(TraversableLike.scala:234) at 
 scala.collection.TraversableLike.map$(TraversableLike.scala:227) at 
 scala.collection.AbstractTraversable.map(Traversable.scala:104) at 
 com.lightbend.rp.servicediscovery.scaladsl.ServiceLocatorLike.translateResolved(ServiceLocator.scala:166) at 
 com.lightbend.rp.servicediscovery.scaladsl.ServiceLocatorLike.translateResolved$(ServiceLocator.scala:162) at 
 com.lightbend.rp.servicediscovery.scaladsl.ServiceLocator$.translateResolved(ServiceLocator.scala:64) at 
 com.lightbend.rp.servicediscovery.scaladsl.ServiceLocatorLike.$anonfun$doLookup$3(ServiceLocator.scala:287) at  

```

## expectation

resource endpoint is discovered.

## notes

Analysis provided by @kikiya 

```scala
def translateResolved(protocol: Option[String], hostname: String, addressARecord: Dns.Resolved): Seq[Service] =
    targetRuntime match {
      case Some(Kubernetes | Mesos) =>
        val uris =
          addressARecord.ipv4.map(v4 => new URI(protocol.orNull, null, v4.getHostAddress, 0, null, null, null)) ++
            addressARecord.ipv6.map(v6 => new URI(protocol.orNull, null, v6.getHostAddress, 0, null, null, null))

        uris.map(u => Service(hostname, u))

      case None =>
        Seq.empty
    }
```

> the port is being set to `0` by default and this is causing

```
textPayload:  "2018-09-04T16:06:29.793Z [[33mwarn[0m] com.lightbend.lagom.internal.persistence.cassandra.CassandraReadSideSessionProvider$ [sourceThread=application-lagom.persistence.dispatcher-50, akkaTimestamp=16:06:29.792UTC, akkaSource=CassandraReadSideSessionProvider$, sourceActorSystem=application] - Failed to connect to Cassandra and initialize. It will be retried on demand. Caused by: All host(s) tried for query failed (tried: /10.128.0.6:0 (com.datastax.driver.core.exceptions.TransportException: [/10.128.0.6:0] Cannot connect))
```

--- a/service-discovery/src/main/scala/com/lightbend/rp/servicediscovery/scaladsl/ServiceLocator.scala
+++ b/service-discovery/src/main/scala/com/lightbend/rp/servicediscovery/scaladsl/ServiceLocator.scala
@@ -126,6 +126,8 @@ trait ServiceLocatorLike {
 
   def translateProtocol(endpoint: String, srvName: String): Option[String] =
     targetRuntime match {
+      case _ if endpoint == "" => Option.empty
+
       case Some(Kubernetes) =>
         Option(srvName.split('.'))
           .filter(_.length >= 2)
@@ -163,11 +165,10 @@ trait ServiceLocatorLike {
     targetRuntime match {
       case Some(Kubernetes | Mesos) =>
         val uris =
-          addressARecord.ipv4.map(v4 => new URI(protocol.orNull, null, v4.getHostAddress, 0, null, null, null)) ++
-            addressARecord.ipv6.map(v6 => new URI(protocol.orNull, null, v6.getHostAddress, 0, null, null, null))
+          addressARecord.ipv4.map(v4 => new URI(protocol.orNull, v4.getHostAddress, null, null)) ++
+            addressARecord.ipv6.map(v6 => new URI(protocol.orNull, v6.getHostAddress, null, null))
 
         uris.map(u => Service(hostname, u))
-
       case None =>
         Seq.empty
     }
@@ -284,7 +285,7 @@ trait ServiceLocatorLike {
                       .map(_.flatten.toVector)
 
                   case r: Resolved =>
-                    Future.successful(translateResolved(translateProtocol(endpoint, r.name), r.name, r))
+                    Future.successful(translateResolved(translateProtocol(endpoint, r.name), name, r))
 
                 }
             } yield result

--- a/service-discovery/src/test/scala/com/lightbend/rp/servicediscovery/scaladsl/ServiceLocatorSpec.scala
+++ b/service-discovery/src/test/scala/com/lightbend/rp/servicediscovery/scaladsl/ServiceLocatorSpec.scala
@@ -153,6 +153,27 @@ class ServiceLocatorSpec extends TestKit(ActorSystem("service-locator", ServiceL
     }
   }
 
+  test("In Kubernetes, ServiceLocator should perform A record resolution when the name doesn't start with underscore") {
+    val mockDnsResolver = TestProbe()
+    val serviceLocator = createServiceLocator(Kubernetes, mockDnsResolver = Some(mockDnsResolver.ref))
+    val result = serviceLocator.lookup("host1.domain")
+
+    mockDnsResolver.expectMsg(Dns.Resolve("host1.domain"))
+    mockDnsResolver.reply(Dns.Resolved("host1.domain1.", Seq(InetAddress.getByName("10.0.12.5"))))
+
+    inside(result.futureValue) {
+      case Vector(s: Service) =>
+        assert(s.hostname === "host1.domain")
+        assert(Option(s.uri.getScheme) === None)
+        assert(Option(s.uri.getUserInfo) === None)
+        assert(s.uri.getHost === "10.0.12.5")
+        assert(s.uri.getPort === -1)
+        assert(s.uri.getPath === "")
+        assert(Option(s.uri.getQuery) === None)
+        assert(Option(s.uri.getFragment) === None)
+    }
+  }
+
   test("In Kubernetes, ServiceLocator.translateName should translate name with namespace + name + endpoint") {
     val serviceLocator = createServiceLocator(Kubernetes)
     val result = serviceLocator.translateName(Some("chirper"), "friendservice", "api")

